sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -157,8 +157,8 @@ abstract class Optimizer(catalogManager: CatalogManager)
     //   since the other rules might make two separate Unions operators adjacent.
     Batch("Union", Once,
       RemoveNoopOperators,
-      RemoveNoopUnion,
-      CombineUnions) ::
+      CombineUnions,
+      RemoveNoopUnion) ::
```

Comment:
Does the added test in `DataFrameSetOperationsSuite` fail without this reorder? The change itself looks reasonable though.

Reply:
Added test is fine. But there will be an [error](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135486/testReport/org.apache.spark.sql/SQLQueryTestSuite/subquery_in_subquery_in_set_operations_sql/) of "Once strategy's idempotence is broken for batch Union".

It is because `CombineUnions` might combine two or more `Union` and they have redundant children under `Distinct`. So the rule order `RemoveNoopUnion` -> `CombineUnions` is not idempotence for Once strategy.

Reply:
Ah, I see.